File: session06_GrowArray_LinkedList_Stack_Queue/GrowArray/BadGrowArray2018S.cc
class BadGrowArray {
private:
  int size;
  int* data;
  
public:
  BadGrowArray() : size(0), data(nullptr) {}
  void addEnd(int v) { //O(n)
    const int* old = data;
    data = new int[size+1];
    for (int i = 0; i < size; i++) // O(size)
      data[i] = old[i];
    data[size] = v;
    size++;
    delete [] old; // you must show us management of memory
  }
  void addStart(int v) { //O(n)
    const int* old = data;
    data = new int[size+1];
    data[0] = v;
    for (int i = 0; i < size; i++)
      data[i+1] = old[i];
    size++;
    delete [] old; // you must show us management of memory
  }
  void insert(int i, int v) { //O(
    const int* old = data;
    data = new int[size+1];
    for (int j = 0; j < i; i++)
      data[j] = old[j];
    data[i] = v;
    size++;
    for (int j = i+1; j < size; j++)
      data[j] = old[j-1];
    delete [] old;
  }
  void removeEnd() { //O(1)
    size--;
  }
  void removeStart() { //O(n)
    size--;
    for (int i = 0; i < size; i++)
      data[i] = data[i+1];
  }
  
  
};

int main(){
  BadGrowArray b;
  b.addEnd(5);
  const int n = 100;
  for (int i = 0; i < n; i++)
    b.addEnd(i);
}

File: old/GrowArray/BadGrowArray2018S.java
public class BadGrowArray2018S {
    private int[] data;
    public BadGrowArray2018S() {
	data = new int[0];
    }
    public void addEnd(int v) { //O(n)
	int[] old = data;
	data = new int[old.length+1]; //O(n)
	for (int i = 0; i < old.length; i++) //O(n)
	    data[i] = old[i];
	data[old.length] = v;
    }
    public void addStart(int v) { //O(n)
	int[] old = data;
	data = new int[old.length+1];
	for (int i = 0; i < old.length; i++)
	    data[i+1] = old[i];
	data[0] = v;
    }
    public void insert(int i, int v) { //O(n)
	int[] old = data;
	data = new int[old.length+1];
	for (int j = 0; j < i; j++)
	    data[j] = old[j];
	data[i] = v;
	for (int j = i+1; j < old.length; j++)
	    data[j+1] = old[j];
    }
    public void removeEnd() { // O(n)
	int[] old = data;
	data = new int[old.length-1];
	for (int i = 0; i < data.length; i++)
	    data[i] = old[i];
    }
    public void removeStart() {
    }
    public int size() {
	return data.length;
    }
    public boolean isEmpty() {
	return data.length == 0;
    }

}
